app/assets/javascripts/discourse/app/models/user.js

```diff
@@ -1087,7 +1092,10 @@ User.reopenClass(Singleton, {
       }
 
       const store = getOwner(this).lookup("service:store");
-      return store.createRecord("user", userJson);
+      const currentUser = store.createRecord("user", userJson);
+      currentUser.appEvents = getOwner(this).lookup("service:appEvents");
```

Comment:
Is there a reason we need to do this? I believe `appEvents` are [auto injected](https://github.com/discourse/discourse/blob/199d58461b425bef1369f10675bcb51d626b4c45/app/assets/javascripts/discourse/app/pre-initializers/inject-discourse-objects.js#L98).

Reply:
There is the reason, yes. We auto-inject, but `appEvents` is not there during the initialization. We create `currentUser` before those calls to `app.inject()`. But note that moving `currentUser.trackStatus();` to the line after auto-injection won't help, at that moment `appEvents` is still not there. This happening because `app.inject()` sets up the factory and don't touch already created objects. So `currentUser.appEvents === null` in initializer. 

But if you access `currentUser` let's say in a controller, at that moment `currentUser` will have `appEvents`, DI will take care of it when instantiating the controller.

I'll likely drop this line soon, but not sure how to do it better yet.

Reply:
Hm, actually, I've found the way we do it. I've fixed it, see https://github.com/discourse/discourse/pull/17236/commits/d11d9c875720fdb092c207a8772929d6939973af. 

My expectations were that [this code](https://github.com/discourse/discourse/blob/199d58461b425bef1369f10675bcb51d626b4c45/app/assets/javascripts/discourse/app/pre-initializers/inject-discourse-objects.js#L97-L103) shouldn't inject anything to models created using the `model.create()` method (as it's explicitly said in [documentation](https://api.emberjs.com/ember/3.15/classes/Application/methods/inject?anchor=inject)), we often create users this way, we call `User.create()`.  But what was strange is that I saw that those users clearly have `store` and `siteSettings` injected. Turned out we have a special place for injecting stuff in models created using `create()`:
https://github.com/discourse/discourse/blob/d11d9c875720fdb092c207a8772929d6939973af/app/assets/javascripts/discourse/app/models/rest.js#L106-L117

Also looks like here is a special place with injections to models created using our own version of data-store that we use instead of Ember Data:
https://github.com/discourse/discourse/blob/d11d9c875720fdb092c207a8772929d6939973af/app/assets/javascripts/discourse/app/services/store.js#L258-L264

I've added only an injection of `appEvents` that I need for this task for now, but I think we need to inject the same services everywhere. [This code](https://github.com/discourse/discourse/blob/199d58461b425bef1369f10675bcb51d626b4c45/app/assets/javascripts/discourse/app/pre-initializers/inject-discourse-objects.js#L97-L103) makes people think that all those services get injected in models, which isn't the case. Also, three places with auto-injections is way too many, we need the only one, but that already [noted](https://github.com/discourse/discourse/blob/d11d9c875720fdb092c207a8772929d6939973af/app/assets/javascripts/discourse/app/services/store.js#L262) in code comments.

Reply:
> We create currentUser before those calls to app.inject().

O this is definitely surprising and thank you for diving deep into the code here to understand the problem. Your fix looks good to me but perhaps we should open a topic on dev as a TODO? As you've said, three places with auto-injections feels like one too many.